lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -0,0 +1,1009 @@
+  /// replaces them with new partial connects.
+  using LazyValue = llvm::function_ref<Value(ImplicitLocOpBuilder &)>;
+  // NOLINTNEXTLINE(misc-no-recursion)
+  void fixupPartialConnect(ImplicitLocOpBuilder &builder, LazyValue dst,
+                           Type dstNewType, Type dstOldType, LazyValue src,
+                           Type srcNewType, Type srcOldType) {
+    // If the types didn't change, just emit a partial connect.
+    if (dstOldType == dstNewType && srcOldType == srcNewType) {
+      auto dstField = dst(builder);
+      auto srcField = src(builder);
+      builder.create<PartialConnectOp>(dstField, srcField);
+      return;
+    }
+    // Check if they are bundle types.
+    if (auto dstOldBundle = dstOldType.dyn_cast<BundleType>()) {
+      auto dstNewBundle = dstOldType.cast<BundleType>();
+      auto srcNewBundle = dstNewType.cast<BundleType>();
+      auto srcOldBundle = srcOldType.cast<BundleType>();
+      for (auto &pair : llvm::enumerate(dstOldBundle)) {
+        // Find a matching field in the old type.
+        auto dstField = pair.value();
+        auto maybeIndex = srcOldBundle.getElementIndex(dstField.name);
+        if (!maybeIndex)
+          continue;
+        auto dstIndex = pair.index();
+        auto srcIndex = *maybeIndex;
+        // Recurse on the matching field. The code is complicated because we are
+        // trying to avoid creating subfield operations when no field ultimately
+        // matches.
+        Value dstValue;
+        auto dstLazy = [&](ImplicitLocOpBuilder &builder) {
+          if (!dstValue)
+            dstValue = builder.create<SubfieldOp>(dst(builder), dstIndex);
+          return dstValue;
+        };
+        auto dstNewElement = dstNewBundle.getElementType(dstIndex);
+        auto dstOldElement = dstOldBundle.getElementType(dstIndex);
+        Value srcValue;
+        auto srcLazy = [&](ImplicitLocOpBuilder &builder) {
+          if (!srcValue)
+            srcValue = builder.create<SubfieldOp>(src(builder), srcIndex);
+          return srcValue;
+        };
+        auto srcNewElement = srcNewBundle.getElementType(srcIndex);
+        auto srcOldElement = srcOldBundle.getElementType(srcIndex);
+        if (!dstField.isFlip)
+          fixupPartialConnect(builder, dstLazy, dstNewElement, dstOldElement,
+                              srcLazy, srcNewElement, srcOldElement);
+        else
+          fixupPartialConnect(builder, srcLazy, srcNewElement, srcOldElement,
+                              dstLazy, dstNewElement, dstOldElement);
+      }
+      return;
+    }
+    // If its not a bundle type, just replace it with a partial connect.
+    builder.create<PartialConnectOp>(dst(builder), src(builder));
+  }
+
+  /// When we replace a bundle type with a similar bundle with different field
+  /// names, we have to rewrite all the code to use the new field names. This
+  /// mostly affects subfield result types and any bulk connects.
```

Comment:
Bitcast and a canonicalization rule?  (sub "foo" (bitcast srcStruct)) -> sub ("bar" srcStruct).